pjlib/src/pj/ssl_sock_apple.m

```diff
@@ -0,0 +1,2013 @@
+/*
+ * Copyright (C) 2019-2020 Teluu Inc. (http://www.teluu.com)
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#include <pj/ssl_sock.h>
+#include <pj/os.h>
+#include <pj/pool.h>
+#include <pj/rand.h>
+
+/* Only build when PJ_HAS_SSL_SOCK and the implementation is Apple SSL. */
+#if defined(PJ_HAS_SSL_SOCK) && PJ_HAS_SSL_SOCK != 0 && \
+    (PJ_SSL_SOCK_IMP == PJ_SSL_SOCK_IMP_APPLE)
+
+#define THIS_FILE               "ssl_sock_apple.c"
+
+#define SSL_SOCK_IMP_USE_CIRC_BUF
+#define SSL_SOCK_IMP_USE_OWN_NETWORK
+
+#include "ssl_sock_imp_common.h"
+#include "ssl_sock_imp_common.c"
+
+#include "TargetConditionals.h"
+
+#include <err.h>
+#include <CoreFoundation/CoreFoundation.h>
+#include <CommonCrypto/CommonDigest.h> 
+#include <Foundation/NSLock.h>
+#include <Network/Network.h>
+#include <Security/Security.h>
+
+/* IMPORTANT note from Apple's Concurrency Programming Guide doc:
+ * "Because Grand Central Dispatch manages the relationship between the tasks
+ * you provide and the threads on which those tasks run, you should generally
+ * avoid calling POSIX thread routines from your task code"
+ * So make sure we don't call any PJLIB functions, not even pj_log(), inside
+ * any dispatch function block.
```

Comment:
That's correct.
- What's the other configuration?
- ok

We have created our own gcd queue:
```assock->queue = dispatch_queue_create("ssl_queue", DISPATCH_QUEUE_SERIAL);```
and I have tried registering GCD threads to PJLIB (that's actually what I did in the first version of this implementation), but there are a couple of drawbacks:
- From [Apple's Concurrency doc](https://developer.apple.com/library/archive/documentation/General/Conceptual/ConcurrencyProgrammingGuide/ThreadMigration/ThreadMigration.html)
Section: `Compatibility with POSIX Threads`
`The underlying thread used to execute a given block can change from invocation to invocation.`
So the threads can keep changing (have tested it here previously and it does change and if I tried to register only at the beginning, after some time, the assertion: `Calling pjlib from unknown/external thread.` will appear. But this is not a big deal of course, since we can simply reregister again (like in sound device callbacks).
- The potential complexity of the network event callback. Calling network event directly from gcd thread, after registering it, did work to some extent, and it passed simple ssl socket tests. But once the callback goes further up to the upper layer and to the application, we have no control over it, and it causes all sorts of problems. This is because as the doc said: `your application should not rely on the following functions returning predictable results between invocations of your block`, which we can't enforce once the callback leaves the ssl socket (for example: ssl_socket's on_connect_complete()->sip_transport_tls->on_transport_state()->...)
- I have a similar idea, using pj_ioqueue_poll(), which will call network_poll() to dispatch the network events. Both approaches (using timer/ioq poll) should give better concurrency performance, but it will be much more complicated of course since we need to handle concurrent events to make sure that the events, and the data read/received, do not get lost (perhaps similar to our current ioqueue_* implementation).

Reply:
> That's correct.
> 
> * What's the other configuration?

actually it is supposed to be "configurable number of threads".

> * ok
> 
> We have created our own gcd queue:
> `assock->queue = dispatch_queue_create("ssl_queue", DISPATCH_QUEUE_SERIAL);`

okay, then is it possible to manage threads executing that queue (e.g: using our own created threads)? So we can register the executing thread. And further, perhaps all SSL sockets can share N threads (instead of N thread(s) for each SSL socket).

> and I have tried registering GCD threads to PJLIB (that's actually what I did in the first version of this implementation), but there are a couple of drawbacks:
> 
> * From [Apple's Concurrency doc](https://developer.apple.com/library/archive/documentation/General/Conceptual/ConcurrencyProgrammingGuide/ThreadMigration/ThreadMigration.html)
>   Section: `Compatibility with POSIX Threads`
>   `The underlying thread used to execute a given block can change from invocation to invocation.`
>   So the threads can keep changing (have tested it here previously and it does change and if I tried to register only at the beginning, after some time, the assertion: `Calling pjlib from unknown/external thread.` will appear. But this is not a big deal of course, since we can simply reregister again (like in sound device callbacks).

Yes, register all of them :) I read somewhere that the number of threads is equal to processor cores number, so it is quite limited.

> * The potential complexity of the network event callback. Calling network event directly from gcd thread, after registering it, did work to some extent, and it passed simple ssl socket tests. But once the callback goes further up to the upper layer and to the application, we have no control over it, and it causes all sorts of problems. This is because as the doc said: `your application should not rely on the following functions returning predictable results between invocations of your block`, which we can't enforce once the callback leaves the ssl socket (for example: ssl_socket's on_connect_complete()->sip_transport_tls->on_transport_state()->...)

IMO on_transport_state() execution should be still inside on_connect_complete() block. In other word, pthread_self(), pthread_getspecific(), etc should return same value in any level of the chain of ssl_socket's on_connect_complete()->sip_transport_tls->on_transport_state()->... Only in another on_connect_complete() that pthread_*() may return different values.

> 
> * I have a similar idea, using pj_ioqueue_poll(), which will call network_poll() to dispatch the network events. Both approaches (using timer/ioq poll) should give better concurrency performance, but it will be much more complicated of course since we need to handle concurrent events to make sure that the events, and the data read/received, do not get lost (perhaps similar to our current ioqueue_* implementation).

Forwarding events to worker thread which involving context switch generally will degrade the performance. And in this code, currently there is only one worker thread, which may degrade the performance further.

Ideally we can invoke the callbacks from the original GCD thread context itself. The problem is that the thread needs to be registered to PJLIB.

Reply:
> actually it is supposed to be "configurable number of threads".

ok. I suppose this configuration will be useful if it can be shared among the SSL sockets, otherwise having N threads per socket seems like an overkill. But then, if we can use gcd, or our existing worker threads (via timer/ioqueue polls) (as we will discuss below), this won't be necessary. So will revisit this later if needed.

> okay, then is it possible to manage threads executing that queue (e.g: using our own created threads)? So we can register the executing thread. And further, perhaps all SSL sockets can share N threads (instead of N thread(s) for each SSL socket).

Not that I know of. The dispatch queue, as per the doc, is supposed to have these advantages, among them:
```
- It eliminates the code needed to create and configure your threads.
- It eliminates the code needed to manage and schedule work on threads.
```
So, by using the dispatch queue, creating threads is supposed to be not necessary anymore.

> Yes, register all of them :) I read somewhere that the number of threads is equal to processor cores number, so it is quite limited.
> Ideally we can invoke the callbacks from the original GCD thread context itself. The problem is that the thread needs to be registered to PJLIB.

Unfortunately, thread registration itself can be a problem. The doc says:
```
Although it is alright to modify the state of a thread while your task is running, you must return the thread to its original state before your task returns. Therefore, it is safe to call the following functions as long as you return the thread to its original state:
pthread_setspecific
```

So since pj_thread_register() calls `pthread_setspecific`, it violates the condition above (and note that the doc also says "must" instead of a more lenient should).
This is exactly the problem that happened in the past, in issue #1837, specifically commit https://github.com/pjsip/pjproject/commit/519b86d54adcea669e9f44befb23b9523bc93fea . Notice that in that commit, the only PJSIP API called is pj_thread_register(), and it caused the crash. And once it's commented or taken out, the crash was gone.

I suppose we have several options now:
- Using gcd
After thinking more about it, the problem is our thread registration requirement and that makes me wonder whether we really really need such requirement. Why do no other libraries have this requirement? Won't it be great if we never encounter that dreaded assertion "Calling pjlib from unknown/external thread" anymore? What features won't we have if we don't require to register the threads?
After some checking, it seems possible to remove the requirement. We will be missing PJ_CHECK_STACK(), which doesn't seem to be crucial nowadays, and which can also be disabled via `PJ_OS_HAS_CHECK_STACK` macro. Another thing is pj_thread_this(), which is more complicated to get around, but seems doable. But maybe I'm missing something here.
So, if we can find a way to disable thread registration (at least on Darwin), we can use gcd, otherwise I'm afraid we have to use the next options.

- Using our existing worker threads.
Timer scheduling, despite having a lite implementation without stack checking, seems to still require lock, which requires thread registration, so it doesn't seem possible.
So perhaps piggybacking ioqueue_poll() seems to be the remaining option? We can use something similar to the implementation in ioqueue_select.c, to check all the network events and process it.

- Using current method in this PR (internal worker thread in the SSL socket)
It's not ideal, I know, but it's more than enough for client use, especially on iOS, which is why this feature is implemented in the first place (i.e. the task item itself is called "iOS native SSL" as per the requests from our users).  So even though it works on Mac as well, it seems secondary, and I'm not sure if there's any demand yet for an SSL server, deployed on Mac (instead of a more popular choice Linux), using native SSL (instead of a more proven option OpenSSL).

Reply:
- GCD requires app to return the thread state, but unfortunately it does not clearly define 'thread state'. If 'thread state' includes thread local storage (TLS), then bigger effort (than working around the thread registration problem) will be needed here, e.g: PJLIB exception also uses TLS and our SIP parser uses exception. So perhaps let's just skip GCD for now.
- Using existing worker thread & piggybacking `ioqueue_poll()`, if it is not so complex to implement, I vote for it, otherwise let's go with current method (which only working for client app, and I guess we need to say it explicitly in the doc).

Reply:
- Since the doc does list the functions that need to revert the thread state, I believe it means whatever state it was before calling those functions.
Right, I missed PJLIB exception.
- My idea was:
   * Add API pj_ioquque_register_poll_cb() to register a function which will be called every time pj_ioqueue_poll() is called.
   * In SSL socket, there's a singleton network event handler (one instance for all sockets) which will put all the network events in a queue. Then when the polling callback is called, it will forward those events to the sockets.

Does this sound okay?

Reply:
Not really sure about formalising the piggyback (`pj_ioquque_register_poll_cb`), IIRC no demand on this so far and better to leave polling function as light as possible (the cost of an additional check may be very2 small, but with the frequency of polling, it will be multiplied). So, what if just hardcode the call to SSL event handler function (macro protected), in ioqueue select & epoll?

Talking about additional cost in polling, if it is possible somehow (GCD issue in our mutex functions), better to use the existing ioqueue mutex for SSL event queue to avoid multiple lock/unlock operations in the poll. If it's not possible, perhaps we can employ a simple integer variable representing pending SSL event number, so only lock the SSL event queue mutex when there is any SSL event (expecting simple integer check to be an atomic operation, perhaps need a specific var attr such as mutable/register).

Reply:
I believe epoll is not available on Mac, so I guess only in ion select.

ioqueue->lock is opaque, so it's inaccessible. Besides, yes, we won't be able to use it anyway because of the GCD issue.

My plan was to use Apple's native objects (instead of PJLIB/PJSIP objects) for this purpose, such as NSLock (see the current implementation which already uses this).

Reply:
Just realised when about to implement it.

Without adding a new API, all ioqueue instances will call ssl_network_poll(), instead of just the one supplied in ssl_sock->param.ioqueue. Is that okay?

Reply:
Good point. I think It should be okay though, as only MacOS apps will be affected. Even if it does not use SSL at all, e.g: media server, the SSL macro protection should remove the function call, right? So only server apps using mix SSL & non-SSL ioqueues will 'suffer'.